### PR TITLE
Cache kernel build artifacts and perform incremental kernel builds

### PR DESCRIPTION
**Pull request by danielocfb**
This change introduces the means for performing incremental kernel builds in CI in order to decrease overall CI turn-around times. We piggy back (and rely) on the kernel's own make-based build infrastructure for that purpose. Specifically, with https://github.com/libbpf/ci/pull/67 we have set the stage for separating build artifacts from source code. With the change at hand we use this capability in conjunction with the `actions/cache` [0] GitHub Actions action to share intermediate build artifacts between CI runs, enabling the kernel build to only rebuild what has changed, as opposed to everything unconditionally.

There are several wrinkles to doing so. The first one is that GitHub Actions checks out the source code anew every time a workflow is run. Because git does not store file meta data like time stamps, but the kernel's build system relies solely on last-modified time stamps, we need a way to approximate the source code time stamps from the previous build. The way we do that is by first remembering both the time stamp and the SHA-1 at which we built and carry that over to subsequent builds. These subsequent builds then make sure to check out said SHA-1, adjust the last-modified time stamps of the entire tree, and then check out the code as it is meant to be tested by the very workflow run in question.

Second, `actions/cache` imposes several restrictions on cached artifacts: only artifacts produced on certain branches can actually be used (to achieve cache isolation) [1], there is a 10 GiB cache-artifact limit per repository [2], and the API is questionable. To circumvent the isolation-imposed requirements, we use the fact that we are maintaining a base branch against which pull requests are created to our advantage. Specifically, once this branch is updated (i.e., pushed to) we trigger a CI run which builds the kernel (already potentially incrementally) and then stores intermediate build artifacts. Pull requests created against this branch (i.e., every one created by Kernel Patches Daemon) will be able to use those artifacts.

Doing so indirectly also addresses the space constraints: the packed build artifacts are <1GiB in size. Because we only need to keep one set of artifacts for each arch and toolchain combo (currently: three) we don't need to worry about trashing (as a side note, these limitations don't seem to be enforced, certainly not rigorously or in a timely manner; I've had >34GiB of build artifacts cached for several hours without an eviction happening).

Because of the questionable API surface, we have to jump through hoops in order to prevent cache creation for regular pull requests: we remove the KBUILD_OUTPUT contents in a separate step, which causes no cache to be created.

As a result of this introduced infrastructure, we see a significant decrease of the "Build Kernel Image" step, which is part of every CI run (happening three times): on GitHub-hosted runners we get down from somewhere in the vicinity of 20minutes to <2minutes. Note that these results depend on what changes the patches in a series make: if a patch modifies a header included by each .c file then the result will be close to a full rebuild irrespective of this infrastructure. Conversely, if a patch only updates documentation the incremental rebuild can be close to instant. In practice, based on a weekend's worth of observations, we seem to be ending up close to "doing almost nothing" almost all the time than to doing a full rebuild.

Note that incremental builds are not bullet proof and to a somewhat large degree we are at the mercy of them working correctly. However, when we encounter a build failure we first attempt a 'make clean' followed by a full rebuild [4]. This is expected to solve build related problems in all conceivable cases. There is, however, the chance of miscompilation or similar issues introduced as part of buggy incremental recompilation logic. A brief survey within Meta's kernel group suggests that generally developers rely on incremental builds and anecdotal problems were always in the "build fails" realm. #dogfood

Note furthermore that selftest and sample builds are not covered by the incremental build machinery. The main reason for that is that neither of them honors KBUILD_OUTPUT -- a deficiency known to upstream Linux folks but not yet addressed. We could work around that conceptually, but payoff is less for these steps and an upstream fix could get us there without added complexity in the CI itself.

Lastly, it is important to note that, should CI, despite all efforts to the contrary, get stuck because a cache somehow contains corrupted data causing us to fail builds, the GitHub UI allows for the deletion of cache artifacts. If no caches are present, this feature is basically a no-op from an initial build perspective (i.e., a full build will be performed).

[0] https://github.com/actions/cache
[1] https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows#restrictions-for-accessing-a-cache
[2] https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows#usage-limits-and-eviction-policy
[3] https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows#deleting-cache-entries 
[4] https://github.com/libbpf/ci/pull/73

Signed-off-by: Daniel Müller <deso@posteo.net>

### CODE REVIEW

**Comment by danielocfb** [comment]
For what it's worth, I had this patch set (in one shape or another) running over the last couple of days on live Patchwork data here: https://github.com/danielocfb/kernel-patches-bpf. I saw the aforementioned build time decrease (with slight variations, depending on patch contents) throughout the bank on all patches. I haven't spotted any issues.

File: .github/workflows/test.yml
Comment (danielocfb) [reply .github/workflows/test.yml:139]: Good question. It mostly just has not been done. It's a big time sink to get this kind of logic tested on a single repository, because of the incredibly slow iteration times (keep in mind that for testing during normal Pacific Time business hours the CI is typically as busy as it gets, causing runs to take up to several hours). Two repositories certainly don't help.

But, I argue, it's also not that important there: pull request are much less frequent and further spaced apart (in time). That means that the base line has more time to diverge, which naturally causes more compilation. It may also result in caches being evicted by the time a new pull request is created. That, combined with the fact that there are just so few pull requests to begin with, means that there is much less to gain there.

I think we should look into it, if for no other reason than to get these two or so lines removed, but I wouldn't treat it as a priority at this point. Let me know if you disagree.

Comment (danielocfb) [reply .github/workflows/test.yml:139]: Actually...there is an actual hard reason, now that I think about it more, and that is that we don't have a baseline in the `vmtest` repository. Basically, because of the hoops we have to jump through due to restrictions by `actions/cache`, we require the `bpf_base` and `bpf-next_base` branches to be present and kept up-to-date by KPD (I didn't require that initially, but it's the only way I could make everything work). That won't happen in `vmtest` (and I don't foresee that changing). We could presumably get somewhere near that with regularly scheduled builds (which I understand is a GitHub Actions feature), but that will require a bit more work. I will make this more clear in the description/code.

File: .github/workflows/test.yml
Comment (danielocfb) [reply .github/workflows/test.yml:192]: I introduced functions as per your suggestion.

**Review by mykola-lysenko — APPROVED** [review]

**Comment by mykola-lysenko** [comment]
let's wait for the build to finish before merge

**Comment by danielocfb** [comment]
Still blocked by https://github.com/libbpf/ci/pull/73 being merged.

**Pull request merged**
